### PR TITLE
Wifi scan results

### PR DESCRIPTION
**Pull request by gpambrozio**
# What does this implement/fix? 

This adds a text sensor to wifi to report on the latest wifi scan results.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuraiton files to keep working)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1054
  
# Test Environment

- [x] ESP32
- [x] ESP8266
- [ ] Windows
- [x] Mac OS
- [ ] Linux

## Example entry for `config.yaml`:

```yaml
text_sensor:
  - platform: wifi_info
    ip_address:
      name: IP Address
    ssid:
      name: Connected SSID
    bssid:
      name: Connected BSSID
    mac_address:
      name: Mac Address
    scan_results:
      name: Scan Results
```

# Explain your changes

Adds a text sensor that shows the latest wifi scan results. This was useful to me to verify what AP might be better for a given sensor and also to tweak the location of certain APs

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by micronen** [comment]
@gpambrozio
I have a very similar PR opened: [#1580](https://github.com/esphome/esphome/pull/1580) 
The main difference is that your version reports RSSI & SSID, and my version adds BSSID, priority,
Please consider merging both PRs, maybe we can add a configuration to support both formats.

**Review by micronen — APPROVED** [review]

File: esphome/components/wifi_info/wifi_info_text_sensor.h [hunk 1]
```diff
@@ -24,6 +24,35 @@ class IPAddressWiFiInfo : public Component, public text_sensor::TextSensor {
   IPAddress last_ip_;
 };
 
+class ScanResultsWiFiInfo : public Component, public text_sensor::TextSensor {
+ public:
+  void loop() override {
```
Comment (OttoWinter) [inline esphome/components/wifi_info/wifi_info_text_sensor.h]:
```suggestion
  void update() override {
```

This calculation shouldn't happen in `loop()` - it's constructing a long string with many memory allocations. With `loop` that will run 60 times per second.

Comment (gpambrozio) [reply esphome/components/wifi_info/wifi_info_text_sensor.h]: Done. Thanks.

Comment (gpambrozio) [reply esphome/components/wifi_info/wifi_info_text_sensor.h]: Actually, I just tried to update one of my devices and I'm getting an error compiling:
```
In file included from src/esphome/components/wifi_info/wifi_info_text_sensor.cpp:1:0:
src/esphome/components/wifi_info/wifi_info_text_sensor.h:29:8: error: 'void esphome::wifi_info::ScanResultsWiFiInfo::update()' marked 'override', but does not override
   void update() override {
        ^
```
@OttoWinter I believe I didn't understand your suggestion fully. I thought that `update` was a method I could override and only ran in certain situations but guess not.

Comment (oxan) [reply esphome/components/wifi_info/wifi_info_text_sensor.h]: You should change the `ScanResultsWiFiInfo` class to inherit from `PollingComponent`.

Comment (gpambrozio) [reply esphome/components/wifi_info/wifi_info_text_sensor.h]: @oxan Thanks, worked perfectly.
@OttoWinter suggestion applied.

**Review by oxan — APPROVED** [review]

File: esphome/components/wifi_info/text_sensor.py [hunk 2]
```diff
@@ -15,6 +16,9 @@
 IPAddressWiFiInfo = wifi_info_ns.class_(
     "IPAddressWiFiInfo", text_sensor.TextSensor, cg.Component
 )
+ScanResultsWiFiInfo = wifi_info_ns.class_(
+    "ScanResultsWiFiInfo", text_sensor.TextSensor, cg.Component
```
Comment (OttoWinter) [inline esphome/components/wifi_info/text_sensor.py]:
```suggestion
    "ScanResultsWiFiInfo", text_sensor.TextSensor, cg.PollingComponent
```

For the updated base class.

File: esphome/components/wifi_info/text_sensor.py [hunk 3]
```diff
@@ -30,6 +34,11 @@
                 cv.GenerateID(): cv.declare_id(IPAddressWiFiInfo),
             }
         ),
+        cv.Optional(CONF_SCAN_RESULTS): text_sensor.TEXT_SENSOR_SCHEMA.extend(
+            {
+                cv.GenerateID(): cv.declare_id(ScanResultsWiFiInfo),
+            }
+        ),
```
Comment (OttoWinter) [inline esphome/components/wifi_info/text_sensor.py]:
```suggestion
        ).extend(cv.polling_component_schema("60s")),
```

File: esphome/components/wifi_info/text_sensor.py (hunk 2 above)
Comment (gpambrozio) [reply esphome/components/wifi_info/text_sensor.py]: Both suggestions applied, thanks.

**Comment by oxan** [comment]
Thanks!

**Pull request merged**